client/src/api/v2/camera.ts

```diff
@@ -0,0 +1,271 @@
+import { useCallback, useEffect, useState } from 'react';
+import formatBytes from 'utils/formatBytes';
+import { camelCaseToStartCase, capitalise } from 'utils/string';
+import { emit, useChannel } from './socket';
+
+enum CameraDevice {
+  Primary = 'primary',
+  Secondary = 'secondary',
+}
+
+interface CameraOverlayStatus {
+  /** List of available overlays */
+  overlays: string[];
+  /** Name of currently active overlay */
+  active: string;
+}
+
+/**
+ * Use a list of camera overlays
+ *
+ * @param device Camera screen
+ * @returns Status of overlays and function to set active overlay.
+ */
+export function useOverlays(device: CameraDevice) {
+  const [overlays, setOverlays] = useState<CameraOverlayStatus | null>(null);
+
+  const handleData = useCallback(
+    (data) => {
+      const json = JSON.parse(data);
+      if (json.device === device) {
+        setOverlays({
+          active: json.activeOverlay,
+          overlays: json.overlays,
+        });
+      }
+    },
+    [device],
+  );
+  useChannel('push-overlays', handleData);
+
+  useEffect(() => {
+    emit('get-overlays');
+  }, [device]);
+
+  const setActiveOverlay = useCallback(
+    (activeOverlay) => {
+      emit(
+        'set-overlays',
+          JSON.stringify({
+            [device]: activeOverlay,
+          }),
+      );
+    },
+    [device],
+  );
+
+  return { overlays, setActiveOverlay };
+}
+
+/**
+ * Send a message to server to start recording video
+ */
+export function startRecording() {
+  emit('start-camera-recording');
+}
+
+/**
+ * Send a message to server to stop recording video
+ */
+export function stopRecording() {
+  emit('stop-camera-recording');
+}
+
+enum CameraRecordingStatus {
+  Off = 'off',
+  Recording = 'recording',
+  Error = 'error',
+}
+
+interface CameraRecordingStatusPayload {
+  /** Status of the camera's recording */
+  status: CameraRecordingStatus;
+  /** Current duration of recording in minutes, if recording */
+  recordingMinutes?: number;
+  /** Filename of recording, if recording */
+  recordingFile?: string;
+  /** Remaining disk space available for recording, in bytes */
+  diskSpaceRemaining: number;
+  /** Description of any error that has occurred */
+  error?: string;
+  /** Catch-all for any extra data */
+  [propName: string]: any;
+}
+
+interface CameraRecordingStatusFormatted {
+  /** Status of the camera's recording */
+  status?: string;
```

Comment:
There's a lot of types with properties that are "unknown" (we don't know whether it exists) rather than "nullable" (it can or can not exist) simply because the data coming from the socket hasn't been validated. This means that anywhere this type is used, it needs to be validated again. What I find is much cleaner is to exhaustively validate inputs when you get them, and raise errors early. That way once the data is inside, we can use type system to be confident that it is properly formatted. Libraries such as [runtypes](https://github.com/pelotom/runtypes) can help by defining the shape of the interface and the validator all in one go. A great article on this philosophy is [Parse, don't validate](https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/).

Reply:
Had a go at runtypes, it's really great! I'll just point out that I've left `data.ts` without runtypes as the whole sensor status page relies on data possibly being not present. V3 will be written smarter :)